tools/metadata/lib/index.ts

```diff
@@ -0,0 +1,173 @@
+  name: string
+  decimals: number
+  address: string
+  slug: string
+}
+
+interface IdexTokens {
+  [symbol: string]: IdexToken
+}
+
+class TokenMetadata {
+  public tokens: NormalizedToken[]
+  public tokensByAddress: { [address: string]: NormalizedToken }
+  public ready: Promise<NormalizedToken[]>
+
+  public constructor() {
+    this.tokens = []
+    this.tokensByAddress = {}
+    this.ready = this.init()
+  }
+
+  public init = async () => {
+    const normalizeIdexToken = (
+      token: IdexToken & { symbol: string }
+    ): NormalizedToken => {
+      return {
+        name: token.name,
+        address: token.address,
+        decimals: token.decimals,
+        symbol: token.symbol,
+        image: `${TRUST_WALLET_IMAGE_API}/${token.address}/logo.png`,
+      }
+    }
+
+    const normalizeMetamaskToken = (
+      token: MetamaskToken & { address: string }
+    ): NormalizedToken => {
+      if (typeof token.decimals !== 'number') {
+        token.decimals = parseFloat(token.decimals)
+      }
+      return {
+        name: token.name,
+        address: token.address,
+        decimals: token.decimals,
+        symbol: token.symbol,
+        image: `${METAMASK_IMAGE_API}/${token.logo}`,
+      }
+    }
+
+    // normalize metamask data
+    const metamaskTokenHashMap: { [key: string]: boolean } = {}
+    const normalizedMetamaskTokens: NormalizedToken[] = Object.entries(
+      contractMap as MetamaskTokens
+    )
+      .map(([address, data]) => {
+        // keep a hash map to give us O(1) lookup later
+        metamaskTokenHashMap[address.toLowerCase()] = true
+        return { ...data, address: address.toLowerCase() }
+      })
+      .filter(
+        contract => !!contract.erc20 && typeof contract.decimals === 'number'
+      )
+      .map(normalizeMetamaskToken)
+
+    // fetch and normalize idex data
+    const { data }: IdexTokens = await axios.get(IDEX_TOKEN_API)
+    const normalizedIdexTokens: NormalizedToken[] = Object.entries(data)
+      .map(([symbol, token]) => {
+        if (token.address) {
+          return { ...token, symbol }
+        }
+      })
+      .map(normalizeIdexToken)
+
+    // persist all metamask tokens to memory metadata array
+    this.tokens.push(...normalizedMetamaskTokens)
+
+    // persist all non-duplicate idex tokens to metadata
+    normalizedIdexTokens.forEach(idexToken => {
+      if (
+        !Object.prototype.hasOwnProperty.call(
+          metamaskTokenHashMap,
+          idexToken.address
+        )
+      ) {
+        this.tokens.push(idexToken)
+      }
+    })
+
+    this._storeTokensByAddress()
+    return this.tokens
+  }
+
+  // get token objects in an array
+  public getTokens = (): NormalizedToken[] => this.tokens
+
+  // get token objects in an object keyed by address
+  public getTokensByAddress = (): { [address: string]: NormalizedToken } =>
+    this.tokensByAddress
+
+  private _storeTokensByAddress = () => {
+    const { tokensByAddress, tokens } = this
```

Comment:
Can we be consistent whether we deconstruct `this` or not? It looks like you use both styles here. Just a personal preference but I like to not deconstruct objects where it makes sense to know where the variable comes from.

Reply:
Usually I will just deconstruct it if there is more than 1 variable, or I'm going to reference that variable a lot. Typing `this` over and over just creates more surface area for bugs imo. I don't feel that strongly about it though, so I'm find doing whatever

Reply:
Yeah I don't really mind too but in cases where I have a local variable that has the same meaning/context as one from props or an object, I like to see the clear distinction. We can ponder on this though